<a href="https://colab.research.google.com/gist/sundimand/d428629bf680582336a6d519139c189a/multiple-regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Import library yang diperlukan**

In [ ]:
import pandas as pd
import numpy as np
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

**Akses Google drive**

In [ ]:
from google.colab import drive
drive.mount('/content/drive') #mengaktifkan google drive anda

Mounted at /content/drive


**Akses file csv ke dataframe**

In [ ]:
df = pd.read_csv('/content/drive/My Drive/Bahan Ajar/SB Statistics for Business/Stats Mhs/Colab/data stat/data4multreg.csv') #lokasi file csv
df

person   Y  X1  X2
0      A  11   4  10
1      B   5   5   6
2      C   7   3   7
3      D   3   2   4
4      E   4   1   3
5      F  12   7   5
6      G  10   8   8
7      H   4   2   4
8      I   8   7  10
9      J   6   1   3

**Multiple regression dengan sklearn**

In [ ]:
from sklearn import linear_model

In [ ]:
x=df[['X1','X2']].values.reshape(-1,2)

#note: reshape(-1,2) sesuai dengan jumlah variable X ada 2, jika X ada 3 maka menjadi: reshape(-1,3)

x

array([[ 4, 10],
       [ 5,  6],
       [ 3,  7],
       [ 2,  4],
       [ 1,  3],
       [ 7,  5],
       [ 8,  8],
       [ 2,  4],
       [ 7, 10],
       [ 1,  3]])

In [ ]:
y=df['Y'].values.reshape(-1,1)
y

array([[11],
       [ 5],
       [ 7],
       [ 3],
       [ 4],
       [12],
       [10],
       [ 4],
       [ 8],
       [ 6]])

In [ ]:
lm = linear_model.LinearRegression()
model = lm.fit(x,y)

In [ ]:
a = lm.intercept_
a

array([2.55172414])

In [ ]:
b = lm.coef_
b

array([[0.67241379, 0.29310345]])

sehingga persamaan regresi linier bergandanya adalah **y = 0.672X1 + 0.293X2 + 2.552**, artinya setiap peningkatan satu satuan pada variable X1 akan memberikan peningkatan sebesar 0.672 pada variable Y dan setiap peningkatan satu satuan pada variable X2 akan memberikan peningkatan sebesar 0.293 pada variable Y.


In [ ]:
y_prediksi = model.predict(x)

In [ ]:
MSE = mean_squared_error(y, y_prediksi)
MSE

3.9913793103448283

In [ ]:
MAE = mean_absolute_error(y, y_prediksi)
MAE

1.6758620689655175

In [ ]:
R2 = r2_score(y, y_prediksi)
R2

0.5565134099616857

**Multiple regression dengan statsmodels.api**

In [ ]:
import statsmodels.api as sm

In [ ]:
x, y = np.array(x), np.array(y)

In [ ]:
x = sm.add_constant(x)
x

array([[ 1.,  4., 10.],
       [ 1.,  5.,  6.],
       [ 1.,  3.,  7.],
       [ 1.,  2.,  4.],
       [ 1.,  1.,  3.],
       [ 1.,  7.,  5.],
       [ 1.,  8.,  8.],
       [ 1.,  2.,  4.],
       [ 1.,  7., 10.],
       [ 1.,  1.,  3.]])

In [ ]:
model = sm.OLS(y, x)
results = model.fit()
results.summary()

/usr/local/lib/python3.10/dist-packages/scipy/stats/_stats_py.py:1806: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=10
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.557
Model:                            OLS   Adj. R-squared:                  0.430
Method:                 Least Squares   F-statistic:                     4.392
Date:                Tue, 31 Oct 2023   Prob (F-statistic):             0.0581
Time:                        12:17:09   Log-Likelihood:                -21.110
No. Observations:                  10   AIC:                             48.22
Df Residuals:                       7   BIC:                             49.13
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.5517      1.944      1.313      0.231      -2.045       7.148
x1             0.6724      0.407      1.652      0.142      -0.290       1.635
x2             0.2931      0.401      0.732      0.488      -0.654       1.240
==============================================================================
Omnibus:                        1.216   Durbin-Watson:                   2.306
Prob(Omnibus):                  0.544   Jarque-Bera (JB):                0.770
Skew:                           0.313   Prob(JB):                        0.680
Kurtosis:                       1.793   Cond. No.                         20.6
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Project** : dengan data table12, data dari kuesionair untuk UMKM, buat persamaan regresi bagaimana pengaruh Knowledge manajemen (KM) dan Competitive intelligence (CI) pada Strategic Management Performance (SMP), berikan penjelasan anda.

**Uji asumsi klasik regresi linier berganda:**

https://www.kaggle.com/code/shrutimechlearn/step-by-step-assumptions-linear-regression

https://towardsdatascience.com/verifying-the-assumptions-of-linear-regression-in-python-and-r-f4cd2907d4c0
